### PR TITLE
Fix DynamicPartitionOpGPU when running on multiple GPUs

### LINKED ISSUE #42500: tf.dynamic_partition causes crash when using multiple GPUs via tf.distribute.MirroredStrategy
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Ubuntu 18.04.2**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: **N/A**
- TensorFlow installed from (source or binary): **binary**
- TensorFlow version (use command below): **2.3.0**
- Python version: **3.6.9**
- Bazel version (if compiling from source): **N/A**
- GCC/Compiler version (if compiling from source): **N/A**
- CUDA/cuDNN version: **CUDA 10.1 / cuDNN 7.6.1**
- GPU model and memory: **RTX 2080 8GB**

**Describe the current behavior**

The `tf.dynamic_partition` operation crashes when running on multiple GPUs using `tf.distribute.MirroredStrategy`.

**Describe the expected behavior**

The same code, also using `tf.distribute.MirroredStrategy` runs succesfully when limited to a single GPU by setting `CUDA_VISIBLE_DEVICES=0`.

**Standalone code to reproduce the issue**

    import tensorflow as tf

    N = 100
    M = 4

    distribute_strategy = tf.distribute.MirroredStrategy()

    def op():
      data = tf.random.uniform((N,))
      partitions = tf.random.uniform((N,), maxval=M, dtype=tf.int32)
      return tf.dynamic_partition(data, partitions, M)

    distribute_strategy.run(op)

**Other info / logs**

Full output of the above code:

```
2020-08-19 12:05:36.898086: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2020-08-19 12:05:37.828508: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcuda.so.1
2020-08-19 12:05:37.900555: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties: 
pciBusID: 0000:09:00.0 name: GeForce RTX 2080 computeCapability: 7.5
coreClock: 1.83GHz coreCount: 46 deviceMemorySize: 7.79GiB deviceMemoryBandwidth: 417.23GiB/s
2020-08-19 12:05:37.901044: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 1 with properties: 
pciBusID: 0000:42:00.0 name: GeForce RTX 2080 computeCapability: 7.5
coreClock: 1.83GHz coreCount: 46 deviceMemorySize: 7.79GiB deviceMemoryBandwidth: 417.23GiB/s
2020-08-19 12:05:37.901070: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2020-08-19 12:05:37.902404: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcublas.so.10
2020-08-19 12:05:37.903988: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcufft.so.10
2020-08-19 12:05:37.904184: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcurand.so.10
2020-08-19 12:05:37.905511: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcusolver.so.10
2020-08-19 12:05:37.906204: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcusparse.so.10
2020-08-19 12:05:37.908753: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudnn.so.7
2020-08-19 12:05:37.910997: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1858] Adding visible gpu devices: 0, 1
2020-08-19 12:05:37.911427: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-08-19 12:05:37.938447: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 2994045000 Hz
2020-08-19 12:05:37.941540: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x49ac240 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-08-19 12:05:37.941596: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-08-19 12:05:42.073849: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x4a182c0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2020-08-19 12:05:42.073925: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): GeForce RTX 2080, Compute Capability 7.5
2020-08-19 12:05:42.073967: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (1): GeForce RTX 2080, Compute Capability 7.5
2020-08-19 12:05:42.075578: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 0 with properties: 
pciBusID: 0000:09:00.0 name: GeForce RTX 2080 computeCapability: 7.5
coreClock: 1.83GHz coreCount: 46 deviceMemorySize: 7.79GiB deviceMemoryBandwidth: 417.23GiB/s
2020-08-19 12:05:42.076318: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1716] Found device 1 with properties: 
pciBusID: 0000:42:00.0 name: GeForce RTX 2080 computeCapability: 7.5
coreClock: 1.83GHz coreCount: 46 deviceMemorySize: 7.79GiB deviceMemoryBandwidth: 417.23GiB/s
2020-08-19 12:05:42.076367: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2020-08-19 12:05:42.076406: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcublas.so.10
2020-08-19 12:05:42.076433: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcufft.so.10
2020-08-19 12:05:42.076457: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcurand.so.10
2020-08-19 12:05:42.076478: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcusolver.so.10
2020-08-19 12:05:42.076499: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcusparse.so.10
2020-08-19 12:05:42.076524: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudnn.so.7
2020-08-19 12:05:42.079838: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1858] Adding visible gpu devices: 0, 1
2020-08-19 12:05:42.079887: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
2020-08-19 12:05:42.939356: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1257] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-08-19 12:05:42.939406: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1263]      0 1 
2020-08-19 12:05:42.939413: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1276] 0:   N N 
2020-08-19 12:05:42.939417: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1276] 1:   N N 
2020-08-19 12:05:42.941258: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:39] Overriding allow_growth setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
2020-08-19 12:05:42.941298: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1402] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 7252 MB memory) -> physical GPU (device: 0, name: GeForce RTX 2080, pci bus id: 0000:09:00.0, compute capability: 7.5)
2020-08-19 12:05:42.942216: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:39] Overriding allow_growth setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
2020-08-19 12:05:42.942237: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1402] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:1 with 2566 MB memory) -> physical GPU (device: 1, name: GeForce RTX 2080, pci bus id: 0000:42:00.0, compute capability: 7.5)
WARNING:tensorflow:Using MirroredStrategy eagerly has significant overhead currently. We will be working on improving this in the future, but for now please wrap `call_for_each_replica` or `experimental_run` or `experimental_run_v2` inside a tf.function to get the best performance.
2020-08-19 12:05:42.972712: F tensorflow/core/kernels/dynamic_partition_op_gpu.cu.cc:108] Non-OK-status: GpuLaunchKernel(GatherOpKernel<T, int32, true>, config.block_count, config.thread_per_block, 0, d.stream(), params, indices, out, gather_dim_size, indices_size, slice_size, out_size) status: Internal: invalid resource handle
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by drebain**
Previously, a callback in DynamicPartitionOpGPU was running outside the correct GPU context and crashed when using a non-default GPU device. This adds a reactivation of the correct context when the callback runs.

Fixes #42500

### CODE REVIEW

File: tensorflow/core/kernels/BUILD [hunk 1]
```diff
@@ -2514,6 +2514,7 @@ tf_kernel_library(
         ":gather_functor",
         ":gpu_prim_hdrs",
         "//tensorflow/core:framework_internal",
+        "//tensorflow/core:gpu_runtime",
```
Comment (sanjoy) [inline tensorflow/core/kernels/BUILD:2517]: Is this necessary?  `check_numerics_op` does not need this.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/BUILD (hunk 1 above)
Comment (drebain) [reply tensorflow/core/kernels/BUILD]: No, I have removed the header that was requiring that, and forgot to remove this.

**Review by sanjoy — DISMISSED** [review]
lgtm

**Review by mdanatg — DISMISSED** [review]

**Comment by rthadur** [comment]
@drebain can you please fix Ubuntu Sanity build failures ?

**Review by sanjoy — APPROVED** [review]

**Pull request merged**